core/lib/state/src/postgres/mod.rs

```diff
@@ -556,23 +563,32 @@ impl ReadStorage for PostgresStorage<'_> {
             .as_ref()
             .and_then(|caches| caches.factory_deps.get(&hash));
 
-        let result = cached_value.or_else(|| {
+        let value = cached_value.or_else(|| {
             let mut dal = self.connection.storage_web3_dal();
             let value = self
                 .rt_handle
-                .block_on(dal.get_factory_dep_unchecked(hash, self.miniblock_number))
-                .expect("Failed executing `load_factory_dep`");
+                .block_on(dal.get_factory_dep(hash))
+                .expect("Failed executing `load_factory_dep`")
+                .map(|(bytecode, inserted_at)| TimestampedFactoryDep {
+                    bytecode,
+                    inserted_at,
+                });
 
             if let Some(caches) = &self.caches {
                 // If we receive None, we won't cache it.
-                if let Some(dep) = value.clone() {
-                    caches.factory_deps.insert(hash, dep);
+                if let Some(value) = value.clone() {
+                    caches.factory_deps.insert(hash, value);
                 }
             };
 
             value
         });
 
+        let result = value
+            .filter(|timestamped_factory_dep| {
+                timestamped_factory_dep.inserted_at <= self.miniblock_number
+            })
+            .map(|timestamped_factory_dep| timestamped_factory_dep.bytecode);
```

Comment:
I'll admit that `and_then` looked better; for some reason I thought that it'd be possible to shortcut a call using `?`: 

```rust
Some(value.filter(|dep| dep.inserted_at <= self.miniblock_number)?.bytecode)
```

Still possible if you move latency observation above 🙃